crates/net/network/src/peers/manager.rs

```diff
@@ -458,6 +481,106 @@ impl Default for PeersConfig {
                 max_inbound: 30,
             },
             reputation_weights: Default::default(),
+            ban_list: Default::default(),
         }
     }
 }
+
+impl PeersConfig {
+    /// A set of peer_ids and ip addr that we want to never connect to
+    pub fn with_ban_list(mut self, ban_list: BanList) -> Self {
+        self.ban_list = ban_list;
+        self
+    }
+}
+
+/// Configuration for the automatic removal of unwanted peers
+#[derive(Debug, Default)]
+pub struct BanList {
+    /// banned peer ids
+    banned_peers: HashSet<PeerId>,
+    /// banned ips
+    banned_ips: HashSet<IpAddr>,
+}
+
+impl BanList {
+    /// creates a new instance from existing values
+    pub fn new(banned_peers: HashSet<PeerId>, banned_ips: HashSet<IpAddr>) -> Self {
+        Self { banned_ips, banned_peers }
+    }
+
+    /// checks the ban list to see if it contains the given ip
+    #[inline]
+    pub(self) fn banned_ip(&self, ip: &IpAddr) -> bool {
+        self.banned_ips.contains(ip)
+    }
+
+    /// checks the banned list to see if it contains the given peer id
+    #[inline]
+    pub(self) fn banned_peer(&self, peer_id: &PeerId) -> bool {
+        self.banned_peers.contains(peer_id)
+    }
+    /// checks the ban list for the given ip address and peer_id and returns true
+    /// if the address is in the ban list
+    #[inline]
+    pub(self) fn is_banned(&self, ip: &IpAddr, peer_id: &PeerId) -> bool {
+        self.banned_ip(ip) || self.banned_peer(peer_id)
+    }
+}
+
+pub enum InboundConnectionError {
+    ExceedsLimit(usize),
+    IpBanned,
+}
+
+#[cfg(test)]
+mod test {
+    use std::{
+        collections::HashSet,
+        net::{IpAddr, Ipv4Addr, SocketAddr},
+    };
+
+    use reth_primitives::{PeerId, H512};
+
+    use crate::{peers::PeerAction, BanList, PeersConfig};
+
+    use super::PeersManager;
+
+    #[tokio::test]
+    async fn test_discovery_ban_list() {
+        let ip = IpAddr::V4(Ipv4Addr::new(127, 0, 1, 2));
+        let socket_addr = SocketAddr::new(ip, 8008);
+        let ban_list = BanList::new(HashSet::default(), HashSet::from_iter(vec![ip]));
+        let config = PeersConfig::default().with_ban_list(ban_list);
+        let mut peer_manager = PeersManager::new(config);
+        peer_manager.add_discovered_node(H512::default(), socket_addr);
+
+        assert!(peer_manager.peers.is_empty());
+    }
+    #[tokio::test]
+    async fn test_on_pending_ban_list() {
+        let ip = IpAddr::V4(Ipv4Addr::new(127, 0, 1, 2));
+        let socket_addr = SocketAddr::new(ip, 8008);
+        let ban_list = BanList::new(HashSet::default(), HashSet::from_iter(vec![ip]));
+        let config = PeersConfig::default().with_ban_list(ban_list);
+        let mut peer_manager = PeersManager::new(config);
+        let a = peer_manager.on_inbound_pending_session(socket_addr.ip());
+        // because we have no active peers this should be fine for testings
+        assert!(a.is_err());
```

Comment:
Can you assert on the exact error? You can do match -> assert on the right case, and panic on all other cases